app/assets/javascripts/shared/editor_toolbar.js

```diff
@@ -148,78 +135,104 @@ class EditorToolbar {
 
-    if (navigator.userAgent.toLowerCase().indexOf('firefox') > -1) { // firefox
-      $element.val(elementText.slice(0, startIndex) + text + elementText.slice(endIndex));
-      $element.trigger('blur'); // Trigger setHeight
+    if (e.originalEvent !== undefined && shrinkEvents.includes(e.originalEvent.inputType)) {
+      // shrink the text area when content is being removed
+      $(this).css({'height': '1px'});
     }
-    else { // all other browsers
-      document.execCommand('insertText', false, text);
+
+    // expand the textarea to fix the content
+    $(this).css({'height': this.scrollHeight + 2});
+  };
+
+  insert(text) {
+    var cursorInfo = this.$target.cursorInfo(),
+        elementText = this.$target.val();
+
+    this.$target.val(elementText.slice(0, cursorInfo.start) + text + elementText.slice(cursorInfo.end));
+  }
+
+  setCursor(affix, cursorInfo) {
+    var adjustedPrefixLength = affix.prefix.length * affix.selection.split('\n').length,
+        adjustedSuffixLength = affix.suffix.length * affix.selection.split('\n').length;
+
+    // post-injection cursor location
+    if (cursorInfo.hasSelection()) { // text was selected, place cursor after the injected string
+      var position = adjustedPrefixLength + cursorInfo.end + adjustedSuffixLength;
+      this.$target[0].setSelectionRange(position, position);
+    } else { // no text was selected, select injected placeholder text
+      this.$target[0].setSelectionRange(cursorInfo.start + affix.prefix.length, cursorInfo.start + affix.asString().length - affix.suffix.length);
     }
   }
 
-  injectSyntax($element, affix) {
-    var adjustedPrefixLength = affix.prefix.length,
-        adjustedSuffixLength = affix.suffix.length,
-        startIndex = $element[0].selectionStart,
-        endIndex = $element[0].selectionEnd,
-        selectedText = $element.val().substring(startIndex, endIndex);
+  injectSyntax(affix) {
+    this.$target.focus(); // bring focus back to $target from the toolbar
+    var cursorInfo = this.$target.cursorInfo(); // Save the original position
 
-    var markdownText = (startIndex == endIndex) ? affix.asPlaceholder : affix.withSelection(selectedText);
+    this.insert(affix.asString());
+    this.setCursor(affix, cursorInfo);
 
-    adjustedPrefixLength *= selectedText.split('\n').length;
-    adjustedSuffixLength *= selectedText.split('\n').length;
+    // Trigger a change event because javascript manipulation doesn't
+    this.$target.trigger('textchange');
+  }
 
-    $element.focus(); // bring focus back to $element from the toolbar
+  insertImagePlaceholder(index, file) {
+    var affix = this.affixesLibrary('image-placeholder', file.name);
+    this.insert(affix.asString());
 
-    this.replace(markdownText, $element);
+    var position = this.$target.val().indexOf(affix.asString()) + affix.asString().length,
+        cursorInfo = new CursorInfo(position, position, undefined);
```

Comment:
I don't think this line is needed. I don't see this `cursorInfo` variable used anywhere.

Reply:
Also correct. Before I was passing it to the setCursor function but it sometimes have weird results so I no longer use the setCursor function here.